File: csharp/tests/Idkollen.Client.Tests/BankIdSeTests.cs
using System.Net;
using System.Text.Json;
using Idkollen.Client.Models;
using Xunit;

namespace Idkollen.Client.Tests;

public class BankIdSeTests
{
    private static (Idkollen.Client.IdkollenClient Client, MockHttpMessageHandler Handler) Build()
    {
        var handler = new MockHttpMessageHandler();
        var http = new HttpClient(handler);
        var client = new Idkollen.Client.IdkollenClientBuilder("cid", "sec")
            .BaseUrl("https://x.test")
            .HttpClient(http)
            .Build();
        return (client, handler);
    }

    [Fact]
    public async Task AuthPending()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new { status = "PENDING", id = "abc", autoStartToken = "tok" });

        var result = await client.BankIdSe.AuthAsync(new BankIdSeAuthRequest { Ssn = "199001011234" });

        var pending = Assert.IsType<BankIdSePending>(result);
        Assert.Equal("abc", pending.Id);
        Assert.Equal("tok", pending.AutoStartToken);

        using var doc = JsonDocument.Parse(handler.LastRequestBody);
        Assert.Equal("199001011234", doc.RootElement.GetProperty("ssn").GetString());
    }

    [Fact]
    public async Task AuthCompleted()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new
        {
            status = "COMPLETED",
            id = "abc",
            ssn = "199001011234",
            name = "Test User",
            givenName = "Test",
            surname = "User",
        });

        var result = await client.BankIdSe.AuthAsync(new BankIdSeAuthRequest());
        var done = Assert.IsType<BankIdSeCompleted>(result);
        Assert.Equal("Test User", done.Name);
    }

    [Fact]
    public async Task AuthFailed()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new { status = "FAILED", id = "abc", error = "userCancel" });
        var result = await client.BankIdSe.AuthAsync(new BankIdSeAuthRequest());
        var failed = Assert.IsType<BankIdSeFailed>(result);
        Assert.Equal("userCancel", failed.Error);
    }

    [Fact]
    public async Task PhoneAuthPending()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new { status = "PENDING", id = "abc", hintCode = "outstandingTransaction" });

        var result = await client.BankIdSe.PhoneAuthAsync(new BankIdSePhoneAuthRequest("199001011234", "user"));

        Assert.IsType<BankIdSePendingPhone>(result);
    }

    [Fact]
    public async Task Verify()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new
        {
            ssn = "199001011234",
            name = "Test User",
            givenName = "Test",
            surname = "User",
            age = 34,
            verifiedAt = "2026-06-08T10:00:00Z",
        });

        var result = await client.BankIdSe.VerifyAsync(new BankIdSeVerifyRequest("qr"));
        Assert.Equal(34, result.Age);
        Assert.Equal("Test User", result.Name);
    }

    [Fact]
    public async Task CancelAuthSendsDelete()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.NoContent, "");
        await client.BankIdSe.CancelAuthAsync("xyz");
        Assert.Equal(HttpMethod.Delete, handler.LastRequest.Method);
        Assert.Equal("/v3/bankid-se/auth/xyz", handler.LastRequest.RequestUri!.AbsolutePath);
    }
}

File: swift/Tests/IdkollenClientTests/BankIdSeTests.swift
import XCTest
@testable import IdkollenClient

final class BankIdSeTests: XCTestCase {
    var client: IdkollenClient!

    override func setUp() {
        super.setUp()
        MockURLProtocol.reset()
        client = IdkollenClientBuilder(clientId: "cid", clientSecret: "sec")
            .baseUrl("https://x.test")
            .urlSession(.mocked())
            .build()
    }

    func testAuthPending() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "PENDING",
            "id": "abc",
            "autoStartToken": "tok",
        ])

        let result = try await client.bankIdSe.auth(BankIdSeAuthRequest(ssn: "199001011234"))
        guard case .pending(let pending) = result else { return XCTFail("expected pending") }
        XCTAssertEqual(pending.id, "abc")
        XCTAssertEqual(pending.autoStartToken, "tok")

        let body = try JSONSerialization.jsonObject(with: MockURLProtocol.capturedBodies.last!) as! [String: Any]
        XCTAssertEqual(body["ssn"] as? String, "199001011234")
    }

    func testAuthCompleted() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "COMPLETED",
            "id": "abc",
            "ssn": "199001011234",
            "name": "Test User",
            "givenName": "Test",
            "surname": "User",
        ])

        let result = try await client.bankIdSe.auth(BankIdSeAuthRequest())
        guard case .completed(let done) = result else { return XCTFail("expected completed") }
        XCTAssertEqual(done.name, "Test User")
    }

    func testAuthFailed() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "FAILED",
            "id": "abc",
            "error": "userCancel",
        ])

        let result = try await client.bankIdSe.auth(BankIdSeAuthRequest())
        guard case .failed(let failed) = result else { return XCTFail("expected failed") }
        XCTAssertEqual(failed.error, "userCancel")
    }

    func testPhoneAuthPending() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "PENDING",
            "id": "abc",
            "hintCode": "outstandingTransaction",
        ])

        let result = try await client.bankIdSe.phoneAuth(BankIdSePhoneAuthRequest(ssn: "199001011234", callInitiator: "user"))
        guard case .pending = result else { return XCTFail("expected pending") }
    }

    func testVerify() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "ssn": "199001011234",
            "name": "Test User",
            "givenName": "Test",
            "surname": "User",
            "age": 34,
            "verifiedAt": "2026-06-08T10:00:00Z",
        ])

        let result = try await client.bankIdSe.verify(BankIdSeVerifyRequest(qrCode: "qr"))
        XCTAssertEqual(result.age, 34)
        XCTAssertEqual(result.name, "Test User")
    }

    func testCancelAuthSendsDelete() async throws {
        MockURLProtocol.enqueue(status: 204, body: Data())
        try await client.bankIdSe.cancelAuth(id: "xyz")
        let req = MockURLProtocol.capturedRequests.last!
        XCTAssertEqual(req.httpMethod, "DELETE")
        XCTAssertEqual(req.url?.path, "/v3/bankid-se/auth/xyz")
    }
}
